dnsprovider/aws.go

```diff
@@ -0,0 +1,315 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package dnsprovider
+
+import (
+	"fmt"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/google/uuid"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/route53"
+
+	"github.com/kubernetes-incubator/external-dns/endpoint"
+	"github.com/kubernetes-incubator/external-dns/plan"
+)
+
+// Route53API is the subset of the AWS Route53 API that we actually use.  Add methods as required. Signatures must match exactly.
+// mostly taken from: https://github.com/kubernetes/kubernetes/blob/853167624edb6bc0cfdcdfb88e746e178f5db36c/federation/pkg/dnsprovider/providers/aws/route53/stubs/route53api.go
+type Route53API interface {
+	ListResourceRecordSetsPages(input *route53.ListResourceRecordSetsInput, fn func(resp *route53.ListResourceRecordSetsOutput, lastPage bool) (shouldContinue bool)) error
+	ChangeResourceRecordSets(*route53.ChangeResourceRecordSetsInput) (*route53.ChangeResourceRecordSetsOutput, error)
+	ListHostedZonesPages(input *route53.ListHostedZonesInput, fn func(resp *route53.ListHostedZonesOutput, lastPage bool) (shouldContinue bool)) error
+	ListHostedZonesByName(input *route53.ListHostedZonesByNameInput) (*route53.ListHostedZonesByNameOutput, error)
+	CreateHostedZone(*route53.CreateHostedZoneInput) (*route53.CreateHostedZoneOutput, error)
+	DeleteHostedZone(*route53.DeleteHostedZoneInput) (*route53.DeleteHostedZoneOutput, error)
+}
+
+// AWSProvider is an implementation of DNSProvider for AWS Route53.
+type AWSProvider struct {
+	Client Route53API
+	DryRun bool
+}
+
+// Zones returns the list of hosted zones.
+func (p *AWSProvider) Zones() ([]string, error) {
+	zones := []string{}
+
+	f := func(resp *route53.ListHostedZonesOutput, lastPage bool) (shouldContinue bool) {
+		for _, zone := range resp.HostedZones {
+			zones = append(zones, *zone.Name)
+		}
+
+		return true
+	}
+
+	err := p.Client.ListHostedZonesPages(&route53.ListHostedZonesInput{}, f)
+	if err != nil {
+		return zones, err
+	}
+
+	return zones, nil
+}
+
+// Zone returns a single zone given a DNS name.
+func (p *AWSProvider) Zone(dnsName string) (*route53.HostedZone, error) {
+	params := &route53.ListHostedZonesByNameInput{
+		DNSName: aws.String(dnsName),
```

Comment:
I don't know. For the initial version I would like to be more rigorous and not add too much sugar. So the value is passed through as it was defined by the user. On GCP missing the trailing dot would result in error, but so would any invalid name. I would add this "common mistakes sugar" in a second step.